src/main/scala/gemmini/ReservationStation.scala

```diff
@@ -281,91 +269,77 @@ class ReservationStation[T <: Data : Arithmetic, U <: Data, V <: Data](config: G
+        new_entry.opa.bits.overlaps(e.bits.opa.bits)})  // war
+
+      new_entry.deps := ld_deps ++ ex_deps ++ st_deps // TODO: doesn't work when entries per type isnt power of 2
+    }.elsewhen (is_ex) {
+      // raw (after ld) | war (after st) | waw (after ld)
+      val ld_deps = VecInit(entries_ld.map { e => e.valid && e.bits.opa.valid && not_config && (
+        new_entry.opa.bits.overlaps(e.bits.opa.bits) || // waw if preload, raw if compute
+        new_entry.opb.bits.overlaps(e.bits.opa.bits))}) // raw
+
+      val ex_deps = VecInit(entries_ex.map { e => e.valid && !e.bits.issued }) // same q
+
+      val st_deps = VecInit(entries_st.map { e => e.valid && e.bits.opa.valid && not_config && new_entry.opa_is_dst &&
+        new_entry.opa.bits.overlaps(e.bits.opa.bits)})  // war
+
+      new_entry.deps := ld_deps ++ ex_deps ++ st_deps
+    }.otherwise {
+      // raw (after ld/ex)
+      val ld_deps = VecInit(entries_ld.map { e => e.valid && e.bits.opa.valid && not_config &&
+        new_entry.opa.bits.overlaps(e.bits.opa.bits)})  // raw
+
+      val ex_deps = VecInit(entries_ex.map { e => e.valid && e.bits.opa.valid && not_config &&
+        e.bits.opa_is_dst && new_entry.opa.bits.overlaps(e.bits.opa.bits)}) // raw only if ex is preload
+
+      val st_deps = VecInit(entries_st.map { e => e.valid && !e.bits.issued }) // same q
+
+      new_entry.deps := ld_deps ++ ex_deps ++ st_deps
+    }
+    // TODO: evaluate whether to keep same q
 
     new_entry.allocated_at := instructions_allocated
 
     new_entry.complete_on_issue := new_entry.is_config && new_entry.q =/= exq
 
-    val is_full = PopCount(Seq(dst.valid, op1.valid, op2.valid)) > 1.U
-    val full_alloc_id = MuxCase((reservation_station_full_entries-1).U, full_entries.zipWithIndex.map { case (e, i) => !e.valid -> i.U })
-    val partial_alloc_id = MuxCase((reservation_station_partial_entries-1).U, partial_entries.zipWithIndex.map { case (e, i) => !e.valid -> i.U })
-
-    when (!is_full && !partial_entries(partial_alloc_id).valid) {
-      io.alloc.ready := true.B
-      partial_entries(partial_alloc_id).valid := true.B
-      partial_entries(partial_alloc_id).bits := new_entry
-      partial_entries(partial_alloc_id).bits.opb.valid := false.B
-      partial_entries(partial_alloc_id).bits.opb.bits := DontCare
-      new_partial_allocs(partial_alloc_id) := true.B
-    } .elsewhen (!full_entries(full_alloc_id).valid) {
-      io.alloc.ready := true.B
-      full_entries(full_alloc_id).valid := true.B
-      full_entries(full_alloc_id).bits := new_entry
-      new_full_allocs(full_alloc_id) := true.B
-    }
+    Seq(
+      (ldq, entries_ld, new_allocs_oh_ld),
+      (exq, entries_ex, new_allocs_oh_ex),
+      (stq, entries_st, new_allocs_oh_st))
+      .foreach { case (q, entries_type, new_allocs_type) =>
+        when (new_entry.q === q) {
+          val is_full = PopCount(Seq(dst.valid, op1.valid, op2.valid)) > 1.U
```

Comment:
We don't need an `is_full` check anymore.

But just to be safe, let's keep the `is_full` variable and add this assert right underneath it:

```
if (q != exq) { assert(!is_full) }
```

Reply:
Added